### PR TITLE
feat: disable send transactions when insufficient balance

### PR DESCRIPTION
**Pull request by kaspars-gailitis**
## Description

Restricts the users from sending tokens whenever their wallet balance is bellow 0 (Sell case) or the input amount is above wallet balance (trade flow).

## Notice

Before submitting a pull request, please make sure you have answered the following:

- [x] Have you followed the guidelines in our [Contributing]('https://github.com/shapeshift/web/CONTRIBUTING.md) guide?
- [x] Have you checked to ensure there aren't other open [Pull Requests](https://github.com/shapeshift/web/pulls) for the same update/change?

## Pull Request Type

- [ ] :bug: Bug fix (Non-breaking Change: Fixes an issue)
- [ ] :hammer_and_wrench: Chore (Non-breaking Change: Doc updates, pkg upgrades, typos, etc..)
- [x] :nail_care: New Feature (Breaking/Non-breaking Change)

## Issue

Closes: https://github.com/shapeshift/web/issues/736

## Testing

Please outline all testing steps

Sell flow:
1. Pull branch locally and  start the server
2. Under "Your Assets" select an asset which you used to have but currently shows 0 balance
3. Hover over "Sell" button

Trade flow:
1. Pull branch locally and  start the server
2. Input an amount in the Trade Input which is above the amount that is available on the user's wallet for that specific token.
3. Hover over "Trade" button

## Screenshots

Sell attempt whenever the user's wallet is below 0:
sell

Trade attempt whenever the input amount is bigger than the wallet amount:
trade

### CODE REVIEW

File: src/components/AssetHeader/AssetActions.tsx [hunk 1]
```diff
@@ -37,14 +40,26 @@ export const AssetActions = ({ isLoaded, assetId, accountId }: AssetActionProps)
...
+          label={
+            notEnoughCryptoToSell()
+              ? translate('common.insufficientFunds')
+              : translate('common.send')
+          }
+          fontSize='md'
+          px={4}
+          hasArrow
+        >
+          <div>
```
Comment (kaspars-gailitis) [inline src/components/AssetHeader/AssetActions.tsx:53]: The div needs to be here because when the direct child of the tooltip is disabled, then the tooltip is also disabled

**Comment by gomesalexandre** [comment]
A few terminology issues on the PR 🙏🏽 
- `buying transactions` -> `sending token`
- `above the input amount is above wallet balance (trade flow).` -> `the input amount is above wallet balance (trade flow).`

Testing this now

File: src/components/AssetHeader/AssetActions.tsx [hunk 2]
```diff
@@ -29,6 +31,7 @@ export const AssetActions = ({ isLoaded, assetId, accountId }: AssetActionProps)
     isConnected ? send.open({ asset: asset, accountId }) : handleWalletModalOpen()
   const handleReceiveClick = () =>
     isConnected ? receive.open({ asset: asset, accountId }) : handleWalletModalOpen()
+  const notEnoughCryptoToSell = (): boolean => bnOrZero(cryptoBalance).eq(0)
```
Comment (gomesalexandre) [inline src/components/AssetHeader/AssetActions.tsx:34]: This shouldn't be a function, react components inherently react (😛) on props changes so if the concern is that it wouldn't react on it if a variable is used, no worries!

Also regarding good practices on naming and this specific terminology:
- checks evaluating to a bool are usually prefixed with `is`/`has`
- the "sell" terminology is wrong, this is about sending which is different

What about:

```javascript
const hasValidBalance = bnOrZero(cryptoBalance).gte(0)
```

And then inverting the happy and unhappy paths in l.45-47 to have the send copy on the left side and insufficient funds one on the right side of the ternary?

File: src/components/Trade/TradeInput.tsx [hunk 3]
```diff
@@ -49,6 +59,12 @@ export const TradeInput = ({ history }: RouterProps) => {
     state: { wallet }
   } = useWallet()
 
+  const cryptoBalance = useAppSelector(state =>
+    selectPortfolioCryptoHumanBalanceByAssetId(state, buyAsset?.currency?.caip19)
+  )
+  const enoughCryptoToSell = (): boolean =>
+    bnOrZero(cryptoBalance).isGreaterThan(buyAsset?.amount || 0)
```
Comment (gomesalexandre) [inline src/components/Trade/TradeInput.tsx:66]: Use `gt/gte` for consistency across the app

File: src/components/AssetHeader/AssetActions.tsx (hunk 2 above)
Comment (kaspars-gailitis) [reply src/components/AssetHeader/AssetActions.tsx:34]: Sure sounds good, will adjust in the `TradeInput` as well. But `isPositive()` won't work. It will return `true` when the value is `0`. I can alternatively call the `isZero` method and inverse that result.
Screenshot 2022-02-01 at 09 40 49

File: src/components/Trade/TradeInput.tsx [hunk 4]
```diff
@@ -49,6 +59,12 @@ export const TradeInput = ({ history }: RouterProps) => {
     state: { wallet }
   } = useWallet()
 
+  const cryptoBalance = useAppSelector(state =>
+    selectPortfolioCryptoHumanBalanceByAssetId(state, buyAsset?.currency?.caip19)
+  )
+  const enoughCryptoToSell = (): boolean =>
```
Comment (gomesalexandre) [inline src/components/Trade/TradeInput.tsx:65]: Same thing here, selectors are reactive by nature so you shouldn't need to be using a function. I'd use a variable (with the same `hasValidBalance` probably) in the render method. Or a derived selector here but that's probably overkill

File: src/components/AssetHeader/AssetActions.tsx (hunk 2 above)
Comment (kaspars-gailitis) [reply src/components/AssetHeader/AssetActions.tsx:34]: but even with `isZero` it will return `false` for negative numbers, so there needs to be a comparison with 0. 🤔

Comment (gomesalexandre) [reply src/components/AssetHeader/AssetActions.tsx:34]: Ohhh nice one lol. So `gte(0)` it is then, you should never have a negative number in this case but my OCD will kick out knowing this could happen :D

Comment (kaspars-gailitis) [reply src/components/AssetHeader/AssetActions.tsx]: yup same, better be extra safe and also make sure the impossible cases don't break things :D

**Review by gomesalexandre — CHANGES_REQUESTED** [review]
Nice one on also implementing the disable on trade! A few minor comments on naming/improvements on the structures to use.

One thing I've noted though is that the `Preview Trade` is loading while rates are calculating, before being disabled. It should obviously be disabled from the exact moment you have an invalid amount (greater than your balance) typed. The only thing I'm not sure is, should we still fetch the rate?

@0xean does it still make sense to fetch and calculate the rate even if the trade is disabled? User could possibly be using the app to quickly check out rates, so I think the best way is to calculate the rate still for informative purposes, but instantly disable the button on impossible trade reactive to user input.

So basically:

1. Instantly disable preview trade button on invalid amount
2. To be checked: Should we still calculate the rate? What about loading state? 

I would propose to keep the rate calculation and keep the button loading state for it, but just have it disabled in case of invalid balance.

**Comment by kaspars-gailitis** [comment]
@gomesalexandre Thanks for the speedy review 🙌 . Made the adjustments. And my two cents on the loading thing as a first time user is to keep the loading state and then disable the button if the balance is not sufficient. 

My first instinct as a user was that the loading state also checks the balance and the responsiveness of changing the amount and initiating loading reassured me that something is happening and the outcome of that is communicated. So from UX perspective I think it is good 👍

**Comment by gomesalexandre** [comment]
Also, it looks like using `Max` produces a `An error occurred getting a quote for your trade.` toast. This is because the quote API throws and [`getZrxSendMaxAmount`](https://github.com/shapeshift/lib/blob/main/packages/swapper/src/swappers/zrx/getZrxSendMaxAmount/getZrxSendMaxAmount.ts) throws with a `No balance found for <token name>`

**Comment by gomesalexandre** [comment]
> @gomesalexandre Thanks for the speedy review 🙌 . Made the adjustments. And my two cents on the loading thing as a first time user is to keep the loading state and then disable the button if the balance is not sufficient.

Np :)! Agreed and that's the same rationale I had in mind - checked other DApps for comparison and yeah it looks like letting the user check for rates even though he has no balance is the way to go!

File: src/components/Trade/TradeInput.tsx [hunk 5]
```diff
@@ -49,6 +59,11 @@ export const TradeInput = ({ history }: RouterProps) => {
     state: { wallet }
   } = useWallet()
 
+  const cryptoBalance = useAppSelector(state =>
+    selectPortfolioCryptoHumanBalanceByAssetId(state, buyAsset?.currency?.caip19)
```
Comment (gomesalexandre) [inline src/components/Trade/TradeInput.tsx:63]:
```javascript
  const sellAssetBalance = useAppSelector(state =>
    selectPortfolioCryptoHumanBalanceByAssetId(state, sellAsset?.currency?.caip19)
  )
  const hasValidBalance = bnOrZero(sellAssetBalance).gt(0)
```

That'll do the trick. Checking for valid balance for swap has a few more ramifications to it (checking the number according to the rate and substracting gas for tx) and isn't part of the ACC so it is fine not to tackle it here!

Comment (kaspars-gailitis) [reply src/components/Trade/TradeInput.tsx]: Nice catch 🙌 . Yea I was also thinking about including the gas in the comparison, but that is a bit outside of this issue.

**Comment by 0xApotheosis** [comment]
In addition to @gomesalexandre's excellent comments, we'll need to update the PR title to [Conventional Commit](https://www.conventionalcommits.org/en/v1.0.0/) standard - maybe something like `feat: disable send transactions when insufficient balance`.

File: src/components/Trade/TradeInput.tsx [hunk 6]
```diff
@@ -278,22 +293,34 @@ export const TradeInput = ({ history }: RouterProps) => {
...
-                wordWrap: 'break-word'
-              }}
+            <Tooltip
+              label={translate('common.insufficientFunds')}
+              fontSize='md'
+              px={4}
+              hasArrow
+              isDisabled={
+                !isValid || isSubmitting || isSendMaxLoading || !!action || hasValidBalance
+              }
```
Comment (gomesalexandre) [inline src/components/Trade/TradeInput.tsx:303]: Non-blocking, but I think it would make sense to use red errored state submit button with insufficient funds copy vs. a tooltip, for consistency across the app.

**Review by gomesalexandre — COMMENTED** [review]
Disable send max button on empty sell asset balance and we should be good to go! Thank you for the quick amends and the work on this 🥇 

Added a non-blocking comment regarding using errored state vs. a tooltip for the trade flow, that should also make the diff a bit smaller.

Also since a follow-up will be created for insufficient trade balance, it would make sense for the title to have something like `empty user balance` vs. `insufficient funds`

**Comment by kaspars-gailitis** [comment]
> Disable send max button on empty sell asset balance and we should be good to go! Thank you for the quick amends and the work on this 🥇

Super fun working on this, thanks for taking the chance on me 🤩. 

> Added a non-blocking comment regarding using errored state vs. a tooltip for the trade flow, that should also make the diff a bit smaller.
> 
> Also since a follow-up will be created for insufficient trade balance, it would make sense for the title to have something like `empty user balance` vs. `insufficient funds`

Removed the tooltip and updated to compare the sell asset balance and amount (doesn't take into consideration gas fees), but should get the job done until the  next changes will take that into consideration.

File: src/components/Trade/TradeInput.tsx [hunk 7]
```diff
@@ -282,16 +290,22 @@ export const TradeInput = ({ history }: RouterProps) => {
...
                 wordWrap: 'break-word'
               }}
             >
               <Text
-                translation={!wallet ? 'common.connectWallet' : error ?? 'trade.previewTrade'}
+                translation={
+                  !wallet
+                    ? 'common.connectWallet'
+                    : isValid && !hasValidTradeBalance
+                    ? 'common.insufficientFunds'
```
Comment (gomesalexandre) [inline src/components/Trade/TradeInput.tsx:306]: Very last one ;) avoid nested ternaries!

**Review by gomesalexandre — APPROVED** [review]
LGTM now! Well done 🎉  
Created a follow-up for calculating gas fees for insufficient balance flow: https://github.com/shapeshift/web/issues/883

Added a last comment but it's not blocking approval, just make sure to make it a separate commit (not an amend) when updating it so it's easy to double-check that the flow is correct in the change from nested ternary.

**Review by gomesalexandre — APPROVED** [review]

**Review by 0xdef1cafe — APPROVED** [review]

**Comment by 0xdef1cafe** [comment]
brilliant - thanks for the PR @kaspars-gailitis! hope to see you picking up more bounties
